README.md
# BEM
BEM is a python package that can simulate electrostatics for ion traps using boundary element method.

This package wraps the fastlap BEM (boundary element method aka BIE boundary integral
equations) + FMM (fast multipole method) + Laplace kernel implementation
from [http://www.rle.mit.edu/cpg/research_codes.htm](http://www.rle.mit.edu/cpg/research_codes.htm). The original code is
in /doc/fastlap_fl-2.0-22oct96 for reference.



## Table of contents
- [Installation](#id-1)
- [Workflow](#id-2)
	1. [Creating the trap model in STL format](#id-3)
	2. [Configuring a new job in BEM](#id-4)
	3. [BEM simulation](#id-5)
	4. [Result analysis](#id-6)
- [Berkeley high performance cluster - Savio](#id-7)
- [Savio Basics](#id-8)

## Installation <div id='id-1'/>
To install BEM on your computer, clone the [bem](https://github.com/HaeffnerLab/bem) directory and create a new conda environment. 

```
conda create -n bem39 python=3.9 
```

Activate the environment. You should see the command prompt change to reflect this (e.g. from base to bem39)

```
conda activate bem39
```

Navigate to the bem directory which contains ```setup.py``` and install BEM


```
pip install .
```

To deactivate the environment,

```
conda deactivate
``` 

The package should be compatible with following systems:
- Windows
- macOS
- Linux

Installation on Berkeley high performance cluster (Savio) is similar.

The ```bem39``` conda environment and the bem repository should already be installed in the shared group directory ```/global/home/groups/fc_haeffnerbem/bem39```. To check whether it is installed, run

```
source activate /global/home/groups/fc_haeffnerbem/bem39
```

If BEM is already installed in the group directory, skip this section. If BEM is not installed, navigate to the group directory, clone the [bem](https://github.com/HaeffnerLab/bem) directory and create a new conda environment. 

```
cd /global/home/groups/fc_haeffnerbem
module load python/3.9.12
conda create -p bem39 python=3.9  
```

Activate the environment. You should see the command prompt change to reflect this (e.g. from base to bem39)

```
source activate /global/home/groups/fc_haeffnerbem/bem39
```

Navigate to the bem directory which contains ```setup.py``` and install BEM


```
pip install .
```

To deactivate the environment,

```
source deactivate
``` 

You can also install it in your own directory ```/global/home/users/<your_user_name>/``` on Savio. Clone the [bem](https://github.com/HaeffnerLab/bem) directory to there and create a new conda environment.
```
cd /global/home/users/<your_user_name>/
module load python/3.9.12
conda create -n bem39 python=3.9
```
Install the package
```
source activate bem39
cd bem
pip install .
```


## Workflow <div id='id-2'/>

Each time you run a BEM analysis, the workflow will consist broadly of four distinct tasks:

1. Creating the trap model in STL format
	- Creating a model of the trap
	- Coloring the trap electrodes
	- Exporting the model as an STL
2. Configuring a new job in BEM
	- Importing the STL model into BEM and re-meshing via BEM
	- Check mesh in 3D to make sure all electrodes are imported and meshed properly
	- Configuring simulation parameters in BEM
3. BEM simulation
	- Running the BEM electrostatics simulation to determine the potentials generated by each electrode.
	- The BEM electrostatics can either be run locally (i.e. on your computer) or on a high performance cluster (HPC) compute node (e.g. savio-bigmem with 20 cores and 500 Gb memory). 
4.  Result analysis
	- View the result in 3D, for example, view electric field square of RF electrodes together with the charge distribution to roughly determine whether the precision is enough
	- Subsequent analysis such as multipole/derivative control will be done using PaulTrapAnalysis package

An example code with annotations can be found in ```examples/SimpleTrap/SimpleTrap.ipynb```. It includes three steps: STL remesh, electrostatic simulation, and 3D visualization of the result.
	



### 1. Creating the trap model in STL format <div id='id-3'/>
There are many ways to create trap model in STL format as long as two requirements are satisfied. Firstly, STL should be in binary format. Secondly, the exported STL has attributes that distinguish electrodes from each other. The attribute can be different colors for different electrodes.

Fusion 360 is the most natural choice to design traps, thus making it the most convenient choice to export STL. However it requires that the whole trap model is a single body, which is not easy to achieve for some designs. Inventor is similar to Fusion 360 but does not have the single body requirement.

Below shows the steps to create and export STL in Fusion 360.

#### 1.1 Coordinates

First, create a model of the trap in Fusion 360. It would be better if your model follows optical coordinates (see figure below). In other words, the trap axis in Fusion 360 is better along the z-axis and the vertical direction.

![imgs/coords.jpg](imgs/coords.jpg)
<center><figcaption>There are two standard conventions for coordinates. In CAD coordinates, x is along the trap axis and in optical coordinates z is along the trap axis. **Throughout BEM, we will always use optical coordinates.**</figcaption></center>

Additionally, ensure that your trap model is a single body in Fusion 360. When exporting, each body becomes its own STL and we wish to have the trap as a single STL.

#### 1.2 Coloring Electrodes

In the Fusion 360 design, the trap is a single body. Therefore, we must be able to distinguish electrode surfaces from one another somehow. This is accomplished by coloring the face of each electrode. Later in BEM, we will identify the electrodes with a mapping between the color used in Fusion 360 and the electrode name.

To facilitate this process, a set of predefined colors for BEM have been created. To import these colors into Fusion 360:

1. In the Modify tab, click on Manage Materials to launch the Material Browser
2. In the Materials Browser, select Appearance on the top bar
3. In the bottom bar of the Material Browser, click on the folder icon, select Open Existing Library and load the BEM color library located ```bem/bemColors_lib/bemColors.adsklib```
4. Now, you should see a bemCol folder appear in the left menu. For *each* color, right click and select Add To/Favorites

![i](imgs/colors1.jpg)

Now color each electrode:

1. Select the face you wish to color
2. Open the Appearance tab (press 'a')
3. Using the search bar, find the appropriate BEM color you wish to use (```bemxx```) and drag that color over the face you wish to color.


![i](imgs/colors2.png)
<center><figcaption>Sample colored model. Notice bem1 is the top left electrode.</figcaption></center>


Note: make sure to color all surfaces that you wish to be included in the simulation. All surfaces that are not colored (and subsequently renamed in BEM) will be discarded by BEM.

Note: why are we specifically using these colors? STL files do not support colors. A (not-so-standardly) accepted workaround is to use the Attribute data of each triangle face to store color information. However, the Attribute is only 2 bytes long while a RGB color is 3 bytes long. Therefore, in general, the original colors will be modified when being saved to a STL. ```bemCol``` handles these color issues by defining a between the ```bemxx``` colors imported into Fusion 360 and the colors BEM reads when importing the trap STL.  

#### 1.3 Exporting to STL

Finally, export your model as an STL. In Fusion 360, select the trap body using the menu on the left-hand side and select Save as Mesh. Make sure to select

- Format: binary STL

Unit is not restricted. It can always be rescaled in BEM. Millimeter is chosen as a convention in Fusion 360. In Inventor, it can also be micron.

Optionally, you can refine the mesh density when exporting in Fusion 360. However, we will be re-meshing in later steps, so this is not strictly necessary.

### 2. Configuring a new job in BEM <div id='id-4'/> 

Firstly, we can process STL files locally. The procedure typically includes reading the STL file, renaming electrodes to DCx or RFx, where x is a number, refining the mesh with area density constraint. Then we can export the processed base mesh in VTK format. Choosing VTK as the mesh format makes it very easy to visualize and iterate the mesh parameters interactively. Visualization can be done by calling 
```
mesh.plot()
```
if mesh is a ```Mesh``` object, or by calling
```
Result.view(prefix, '')
```
if you want to directly visualize the mesh VTK file within the same folder.

![i](imgs/base_mesh_3D.png)
<center><figcaption>3D visualization of the processed base mesh. Electrodes are colored differently. The top-left slider can choose different electrode and highlight them with bold edges.</figcaption></center>


Secondly, we can either configure the job in the same Jupyter notebook, or read the mesh with
```
mesh, _ = Mesh.from_vtk(prefix)
```
and write the job in a separate python file. This makes it possible to run the job on high performance cluster after simply uploading the job file (e.g. ```examples/SimpleTrap/run.py```), the mesh file (e.g. ```examples/SimpleTrap/SimpleTrap_mesh.vtk```), as well as the job description file (e.g. ```examples/SimpleTrap/saviojob.sh```). 

Inside the job file, it typically involves reading the base mesh, defining the grid to extract data, creating parallel jobs and execute them.


### 3. BEM simulation <div id='id-5'/> 

Both local and Berkeley's cluster parallel computation can use ipyparallel package, so the code is just the same. In theory, local execution not only can be done in Jupyter notebook, but also can be achieved by running 
```
ipython run.py
``` 
in the terminal. Remote execution on Berkeley HPC (Savio) requires running 
```
sbatch saviojob.sh
```

The data will be saved as VTK files. The file named as ```<prefix>_<electrode_name>.vtk``` stores potential, field (RF only), and field square (RF only). The file named as ```<prefix>_<electrode_name>_mesh.vtk``` stores the charge distribution, and the final mesh after adaptive mesh during simulation.

### 4. Result analysis <div id='id-6'/> 

Lastly, after we get the results, we can visualize the result by simply calling 
```
Result.view(prefix, '<electrode_name>')
```
For example, if we want to visualize the field square of the SimpleTrap's RF electrode, we can simply call ```Result.view(prefix, 'RF')``` in the same folder as the result VTK files. Here is the visualization example

![i](imgs/RF_3D.png)
<center><figcaption>3D visualization of the field square of the RF electrode. Charge distribution and contours of the field square are plotted. The top-right slider can move a free contour around.</figcaption></center>

## Berkeley high performance cluster - Savio<div id='id-7'/>

BRC-HPC is a high-performance computing cluster run by Berkeley Research Computing (BRC). The cluster consists broadly of login nodes and compute partitions. Data and software are stored in the cluster within your personal directory or a *shared group directory*. All nodes (login and compute) have access to these directories.

![](imgs/savio.jpeg)

SSH-ing into the cluster is handled by the 4 login nodes (a random node is assigned). You may test commands and scripts within the login node, but the login nodes are a shared resource and not meant to handle large computational tasks.

To schedule computational tasks (i.e. BEM simulations) to be run on a computing cluster, Savio uses slurm, a job scheduler. This is described in the sections below.

## Savio Basics <div id='id-8'/>

The full user-guide for Savio may be found [here](https://docs-research-it.berkeley.edu/services/high-performance-computing/user-guide/).

Note: if this is the first time you are reading this documentation, I suggest skipping the remainder of this section and returning when you are ready to run BEM on savio and wish to learn the specifics.

### Service Units

Our lab is granted 300,000 yearly service units via Hartmut's faculty allowance. To check that you have access to this computing allowance, login to [MyBRC](https://mybrc.brc.berkeley.edu/user/login) and check that the BRC Cluster Project ```fc_haeffnerbem``` appears on your dashboard. To be added to the project, please contact Hartmut or Yi.

Note: service units are calculated based on the number of cores in the node you requested. E.g., if you run a program for 1 hr on a single core of a 16-core node, you will be charged 16 service units. Therefore, it is in your interest to use as many cores as possible (i.e. as many cores as the available memory allows).

### SSH

To ssh into the a login node, 

```ssh your_BRC_username@hpc.brc.berkeley.edu```

*The password is not your BRC pin*. One-time password authentication via your phone is required. If you have not already, set up OTP via Google Authenticator by following these [directions](https://docs-research-it.berkeley.edu/services/high-performance-computing/user-guide/setting-otp/). The password you must enter is ```XY``` where ```X``` is your BRC pin and immediately following with no spaces is ```Y``` the 6-digit one-time code from Authenticator. 

### Access to the Group Computing Allowance

The hardware configurations for each of the Savio partitions may be found [here](https://docs-research-it.berkeley.edu/services/high-performance-computing/user-guide/hardware-config/#Savio-Hardware). To find which accounts and partitions you have access to,

```sacctmgr -p show associations user=$USER```

![](imgs/sacctmgr.png)

Since our simulations are generally limited by memory, we generally would like to choose bigmem nodes.

### Modules and Custom Software

A number of modules are installed in Savio. These modules can be loaded with the `module` command. For example, every time we run BEM, we want to first load Python (you can run Python without loading the Python module, but it is an outdated version).

`module load python`

**Note: this is a commonly forgotten step**

A final quirk in using conda environments on Savio *not* using the normal `conda activate module_name` but rather,

`source activate module_name`

**This is also a common source of error.**


### Scheduling and Monitoring Jobs with slurm

TODO TALK ABOUT SLURM SBATCH

The ```sq``` module shows recent and pending jobs,

```
module load sq
sq
```


Note: the directory where you run the ```sbatch``` command is the directory where your batch script runs.

Note: you have to deactivate the environment before running the sbatch command

### Transferring Data

Data transfer can be done using Globus. The directions can be found [here](https://docs-research-it.berkeley.edu/services/high-performance-computing/user-guide/data/transferring-data/using-globus-connect-savio/).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## TO DO:
- nothing at the moment

## 2022-10 Shuqi Xu

### Change
- Replace Mayavi with PyVista
- 3D visualization of the mesh supports 28 different colors. If the number of electrodes exceeds 28, duplicate colors will be used.
- A slider is added to highlight the chosen electrode with bold edges. This can be used to identify electrodes.
- Adjust the judgment condition of function ```partition_normals_interception``` inside stl.py file. Now it classify two triangles within the same plane but with opposite normals to the same face.
- Adjust the judgment condition tolerance inside stl.py file. It is now unitless.
- Can read mesh from vtk.
- Also install ```helper_functions``` and ```bemColors```. To import, run ```from helper_functions import helper_functions``` and ```from bem.bemColors_lib.bemColors import bemColors```
- Tested compatible with high performance cluster
- Improve README from Yi's bem-savio

## 2022-07 Wenhao He

### Change
- make this package compatible in Windows
    - main change triangle/triangulate.c and triangle/triangulate.h
    - for windows, we must change long into __int64
    - reference: 	https://blog.csdn.net/cuihaolong/article/details/72842026.  and   	https://www.796t.com/post/YThtcXk=.html
- 7/11/2022: tested compatible with latest vtk==9.0.3, latest mayavi==4.7.4, and old PyQt5==5.15.6 on python==3.9.12 on Linux(Ubuntu 20.04.3) , macOS(Monterey 12.1 intel) , HPC(CentOS Linux release 7.7.1908 (Core)) and Windows(Windows 10 pro 21H2).



## 2022-05 Wenhao He

### Change
- simulate overhang htrap(3D electrode): original code(especially triangulation part) behaves bad for 3D trap, so a lot of changes are made in Bem.
    - unify same points in mesh structure
    - counter clockwise triangle when passed into the package
    - originally, parallel planes will be classified as one plane, which leads to error for 3D cases. Fix now
    - now we can combine triangles first and cut it, so sharp triangles are eliminated
    - fix the error: exceed float precision .......
- a more convenient way to load color (explained in the document)
    - a set of standard color
    - more regular color names
- simplify the environment set up process, with one command: pip install .
    - setup.py, setup.cfg, pyproject.toml and MANIFEST.in
- compatible with python 3.9
    - solve the problem of multiprocessing
    - solve the environment for mayavi PyQt vtk
    - get around mayavi in our workflow
        - our code can run on HPC(CentOs)
        - saves a lot of effort to maintain the code in the future
- others
    - make unit problem clear(micronmeter or millimeter are both ok now)
    - split readme into readme+changelog+copying
    - bugs in ./fastlap
    - other small bugs

- add a document: ./examples/document

## 2022-03 Shuqi Xu

### Change
- fix bugs on partition normal
- fix declaration error on building process
- update documentation
- update cythonize version
- 3/1/2020: tested compatible with latest vtk==9.0.1 and latest mayavi==4.7.2 on python==3.6.12 on both Linux and macOS. Need more test on newer Python versions and on Windows.
- etc












helper_functions/helper_functions.py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib as mpl
from bem.formats import stl
import os
from scipy.signal import argrelextrema
from bem import Result
import pickle
import copy


#trans_file = 'vtk'
trans_file = 'pkl'

def load_file(Mesh,Electrodes,prefix,scale,use_stl=True):
    if not use_stl:
        # load electrode faces from loops
        ele = Electrodes.from_trap(open("%s.ele" % prefix), scale)
        # initial triangulation, area 20, quiet
        mesh = Mesh.from_electrodes(ele)
        mesh.triangulate(opts="a0.01q25.")
    else:
        # load electrode faces from colored stl
        # s_nta is intermediate processed stl file.
        s_nta = stl.read_stl(open("trapstl/%s.stl" % prefix, "rb"))
        mpl.rcParams['lines.linewidth'] = 0.2
        print("Import stl:", os.path.abspath("./trapstl/%s.stl" % prefix), "\n")
        print("Electrode colors (numbers):\n")
        mesh = Mesh.from_mesh(stl.stl_to_mesh(*s_nta, scale=scale / 1e-3, rename={0: "DC21"},print_dropping = True))
    return mesh,s_nta

#Create custom subplot w/ dimensions that you want, add the trapping point,
#then use mesh object's 'plot' function to add the mesh to it.
def plot_mesh(x,y,z,mesh,mesh_unit,title='',fout='fig',save=False,dpi=300,three_d=False):
    '''
    x,y: trapping location (projected along z)
    fout: name of file to save save figure
    '''
    # Plot triangle meshes.
    if not three_d:
        fig, ax = plt.subplots(subplot_kw=dict(aspect="equal"), figsize=(12, 6), dpi=400)
        mesh.plot(ax, three_d)
        ax.set_xlabel(f"z [{np.round(mesh_unit/1e-6)} $\mu$m]", fontsize=10)
        ax.set_ylabel(f"x [{np.round(mesh_unit/1e-6)} $\mu$m]", fontsize=10)
        ax.scatter(z, x, marker='.', color='k',label=f'(z0,x0)=({z},{x})', s=5)
        ax.scatter(0, 0, marker='x', color='r',label='(0,0)', s=5)
        ax.set_title(title)
        ax.legend(loc='center left', bbox_to_anchor=(1, 0.5))
    else:
        fig = plt.figure(figsize=(10,10))
        ax = fig.add_subplot(projection='3d')
        mesh.plot(ax, three_d)
        ax.set_xlabel(f"x [{np.round(mesh_unit/1e-6)} $\mu$m]", fontsize=10)
        ax.set_ylabel(f"y [{np.round(mesh_unit/1e-6)} $\mu$m]", fontsize=10)
        ax.set_zlabel(f"z [{np.round(mesh_unit/1e-6)} $\mu$m]", fontsize=10)
        ax.scatter(0, 0, 0, marker='x', color='r',label='(0,0,0)', s=30)
        ax.scatter(x, y, z, marker='.', color='r',label=f'(x0,y0,z0)=({x},{y},{z})', s=20)      
        ax.legend(loc='center left', bbox_to_anchor=(1, 0.5))
        ax.azim = 300

    if save:
    	plt.savefig(fout, bbox_inches='tight', dpi=dpi)
    plt.show()

# Trap simulations.
def run_job(args):
    # job is Configuration instance.
    job, grid, prefix = args
    
    print('Running job '+job.name)
        
    # refine twice adaptively with increasing number of triangles, min angle 25 deg.
    # job.adapt_mesh(triangles=4e2, opts="q25Q")
    # job.adapt_mesh(triangles=1e4, opts="q25Q")
    # solve for surface charges
    job.solve_singularities(num_mom=4, num_lev=3, max_iter=500)
#     print("done")
    # get potentials and fields
    # For "RF", field=True computes the field.
    result = job.simulate(grid, field=job.name=="RF", num_lev=1)
    
    result.save(prefix,'pkl')
    print("finished job %s" % job.name)
    return job.collect_charges()


def plot_RF(Result,prefix,grid):
    result = Result.load(prefix, "RF",trans_file,trans_file)
    p = result.potential
    maxp = np.amax(p)
    print("p max", maxp)
    x,y,z = grid.to_xyz() # p.shape[0]/2 is in the middle of x.
    p = p[p.shape[0] // 2,:,:] # get a slice of yz plane at x = p.shape[0]/2.
    print("yz plane, RF pseudo")
    fig, ax = plt.subplots()
    fig.set_size_inches(4, 10)
    ax.set_aspect("equal")
    ax.grid(axis='both')
    yticks = np.arange(-0.1, 0.1, 0.01)
    ax.set_yticks(yticks)
    xticks = np.arange(-0.1, 0.1, 0.01)
    ax.set_xticks(xticks)
    # ax.set_ylim(0.5, 1.5)
    # ax.set_xlim(0.5,1.5)
    range = 0.35
    ax.contour(y,z, np.transpose(p), levels=np.linspace((p.max()-p.min())*range+p.min(),(p.max()-p.min())*(1-range)+p.min(), 100), cmap=plt.cm.RdYlGn)
    plt.show()

def plot_DC(Result,prefix,suffix,grid,strs,dir='x'):
    p = np.zeros(Result.load(prefix + suffix, strs[0],trans_file).potential.shape)
    for em in strs:
        ele = em
        result = Result.load(prefix + suffix, ele,trans_file)
        pmid = result.potential
        maxp = np.amax(p)
        #     print("p max", maxp)
        #     print(np.shape(Vx))
        p = p+pmid
    x,y,z = grid.to_xyz()
    if dir== 'x':
        p = p[p.shape[0] // 2,:,:]
        xp = y
        yp = z
    elif dir== 'y':
        p = p[:,p.shape[1] // 2,:]
        xp = x
        yp = z
    else:
        p = p[:,:,p.shape[2] // 2]
        xp = x
        yp = y
    print("yz plane, %s potential" % ele)
    fig, ax = plt.subplots()
    ax.set_aspect("equal")
    # yz plane should use x[1], x[2]. wwc
    X, Y = np.meshgrid(xp, yp)
    fig.set_size_inches(20, 10)
    ax.contour(yp,xp, p, levels=np.linspace(p.min(),p.max(), 20), cmap=plt.cm.RdYlGn)
    plt.show()

def find_saddle(V,X,Y,Z,dim, scale=1, Z0=None,min=False):
    """Returns the indices of the local extremum or saddle point of the scalar A as (Is,Js,Ks).
    V is a 3D matrix containing an electric potential and must solve Laplace's equation
    X,Y,Z are the vectors that define the grid in three directions
    Z0: Z coordinate for saddle finding in a 2D potential slice
    For dim==2, the values of A are linearly extrapolated from [Z0] and [Z0]+1
    to those corresponding to Z0 and Ks is such that z[Ks]<Z0, z[Ks+1]>=Z0."""

    if (dim==2 and Z0==None):
        return 'z0 needed for evaluation'
    if dim==3:
        if len(V.shape)!=3:
            return('Problem with find_saddle.m dimensionalities.')
        # Normalize field
        f=V/float(np.amax(V))
        #next, find the gradient of our normalized potential
        #We have three output fields [Ex,Ey,Ez] from one input potential
        #because a gradient is a vector, so we associate {Ex_i,Ey_i,Ez_i} and f_i
        #for a unique location i in our 'cube' of field locations.
        #I call it 'E' because grad(V)= electric field= E
        [Ex,Ey,Ez]=np.gradient(f,abs(X[1]-X[0])/scale,abs(Y[1]-Y[0])/scale,abs(Z[1]-Z[0])/scale)

        [Ex2,Ey2,Ez2]=np.gradient(f,abs(X[1]-X[0])/scale,abs(Y[1]-Y[0])/scale,abs(Z[1]-Z[0])/scale,edge_order=2)# grid spacing is automatically consistent thanks to BEM-solver

        #return the magntiude of our gradient vector at each field point.
        E=np.sqrt(Ex**2+Ey**2+Ez**2) # magnitude of gradient (E field)

        #This is a little confusing and should be re-written (I added it after calculating the saddle)
        #Setting E=f just means we are going to calculate the global minimum potential pt.
        #Hopefully this is the the trapping location.
        if min== True:
            E=f

        #Next, we begin our search for saddle (if min=False)
        #or, we begin our search for the potential minimum (if min=True)

        #Start the search by looking at the very first location
        m=E[0,0,0]
        #meh
        origin=[0,0,0]

        #Now, we begin searching through our list of field pts (potential or |E-field|)
        for i in range(E.shape[0]):
            for j in range(E.shape[1]):
                for k in range(E.shape[2]):
                    if E[i,j,k]<m:
                        # if Ex2[i,j,k]+Ey2[i,j,k]+Ez2[i,j,k]<0:
                        m=E[i,j,k]
                        origin=[i,j,k]
        #I think this checks if we need to expand our simulation....
        #but looking at looking at saddle identification, I don't think this is actually
        #working b/c the saddle is clearly out of bounds.
        if origin[0]==(0 or V.shape[0]):
            print('find_saddle: Saddle out of bounds in  x (i) direction.\n')
            return origin
        if origin[0]==(0 or V.shape[1]):
            print('find_saddle: Saddle out of bounds in  y (j) direction.\n')
            return origin
        if origin[0]==(0 or V.shape[2]):
            print('find_saddle: Saddle out of bounds in  z (k) direction.\n')
            return origin
    #################################################################################################
    if dim==2: # Extrapolate to the values of A at z0.
        V2=V
        if len(V.shape)==3:
            Ks=0 # in case there is no saddle point
            for i in range(len(Z)):
                if Z[i-1]<Z0 and Z[i]>=Z0:
                    Ks=i-1
                    if Z0<1:
                        Ks+=1
            Vs=V.shape
            if Ks>=len(Z):
                return('The selected coordinate is at the end of range.')
            v1=V[:,:,Ks]
            v2=V[:,:,Ks+1]
            V2=v1+(v2-v1)*(Z0-Z[Ks])/(Z[Ks+1]-Z[Ks])
        V2s=V2.shape
        if len(V2s)!=2: # Old: What is this supposed to check? Matlab code: (size(size(A2),2) ~= 2)
            return('Problem with find_saddle.py dimensionalities. It is {}.'.format(V2s))
        f=V2/float(np.max(abs(V2)))
        [Ex,Ey]=np.gradient(f,abs(X[1]-X[0]),abs(Y[1]-Y[0]))
        E=np.sqrt(Ex**2+Ey**2)
        m=float(np.min(E))
        mr=E[0,0]
        Is,Js=1,1 # in case there is no saddle
        for i in range(E.shape[0]):
            for j in range(E.shape[1]):
                if E[i,j]<mr:
                    mr=E[i,j]
                    Is,Js=i,j
        origin=[Is,Js,Ks]
        if Is==1 or Is==V.shape[0]:
            print('find_saddle: Saddle out of bounds in  x (i) direction.\n')
            return origin
        if Js==1 or Js==V.shape[1]:
            print('find_saddle: Saddle out of bounds in  y (j) direction.\n')
            return origin
    return origin

def find_saddle_drag(V,X,Y,Z,dim, scale=1, Z0=None,min=False):
    """Returns the indices of the local extremum or saddle point of the scalar A as (Is,Js,Ks).
    V is a 3D matrix containing an electric potential and must solve Laplace's equation
    X,Y,Z are the vectors that define the grid in three directions
    Z0: Z coordinate for saddle finding in a 2D potential slice
    For dim==2, the values of A are linearly extrapolated from [Z0] and [Z0]+1
    to those corresponding to Z0 and Ks is such that z[Ks]<Z0, z[Ks+1]>=Z0."""

    if (dim==2 and Z0==None):
        return 'z0 needed for evaluation'
    if dim==3:
        if len(V.shape)!=3:
            return('Problem with find_saddle.m dimensionalities.')
        # Normalize field
        E=V

        #Start the search by looking at the very first location
        m=E[0,0,0]
        #meh
        origin=[0,0,0]
        dragPath = np.zeros((len(E[:,0,0]),2))
        dragVal = np.zeros(len(E[:,0,0]))
        #Now, we begin searching through our list of field pts (potential or |E-field|)
        for i in range(E.shape[0]):
            # if Ex2[i,j,k]+Ey2[i,j,k]+Ez2[i,j,k]<0:
            minval = np.amin(E[i,:,:])
            location = np.where(E[i,:,:] == minval)
            dragPath[i] = location
            dragVal[i] = minval
        #I think this checks if we need to expand our simulation....
        #but looking at looking at saddle identification, I don't think this is actually
        #working b/c the saddle is clearly out of bounds.
        if origin[0]==(0 or V.shape[0]):
            print('find_saddle: Saddle out of bounds in  x (i) direction.\n')
            return origin
        if origin[0]==(0 or V.shape[1]):
            print('find_saddle: Saddle out of bounds in  y (j) direction.\n')
            return origin
        if origin[0]==(0 or V.shape[2]):
            print('find_saddle: Saddle out of bounds in  z (k) direction.\n')
        if min:
            idx = np.where(dragVal == np.min(dragVal))
            absmin = dragPath[idx]
            outlist = np.append(idx, absmin)
        else:
            idx = np.where(dragVal == np.max(dragVal))
            candidates = argrelextrema(dragVal, np.greater)[0]
            if len(candidates) == 0:
                if dragVal[0] < dragVal[len(dragVal)-1]:
                    idx = 0
                else:
                    idx = len(dragVal)
            if len(candidates) == 1:
                idx = candidates[0]
            else:
                for i in range(len(candidates)):
                    if i == 0:
                        idx = candidates[0]
                    elif dragVal[candidates[i]] < dragVal[idx]:
                        idx = candidates[i]
            plt.plot(dragVal)
            plt.plot(idx,dragVal[idx-1],'x')
            plt.plot(len(X)//2,V[len(X) // 2, len(Y) // 2, len(Z) // 2],'o')
            plt.show()
            sad = dragPath[idx-1]
            outlist = np.append(idx,sad)
        return outlist.astype(int)
    #################################################################################################
    if dim==2: # Extrapolate to the values of A at z0.
        V2=V
        if len(V.shape)==3:
            Ks=0 # in case there is no saddle point
            for i in range(len(Z)):
                if Z[i-1]<Z0 and Z[i]>=Z0:
                    Ks=i-1
                    if Z0<1:
                        Ks+=1
            Vs=V.shape
            if Ks>=len(Z):
                return('The selected coordinate is at the end of range.')
            v1=V[:,:,Ks]
            v2=V[:,:,Ks+1]
            V2=v1+(v2-v1)*(Z0-Z[Ks])/(Z[Ks+1]-Z[Ks])
        V2s=V2.shape
        if len(V2s)!=2: # Old: What is this supposed to check? Matlab code: (size(size(A2),2) ~= 2)
            return('Problem with find_saddle.py dimensionalities. It is {}.'.format(V2s))
        f=V2/float(np.max(abs(V2)))
        [Ex,Ey]=np.gradient(f,abs(X[1]-X[0]),abs(Y[1]-Y[0]))
        E=np.sqrt(Ex**2+Ey**2)
        m=float(np.min(E))
        mr=E[0,0]
        Is,Js=1,1 # in case there is no saddle
        for i in range(E.shape[0]):
            for j in range(E.shape[1]):
                if E[i,j]<mr:
                    mr=E[i,j]
                    Is,Js=i,j
        origin=[Is,Js,Ks]
        if Is==1 or Is==V.shape[0]:
            print('find_saddle: Saddle out of bounds in  x (i) direction.\n')
            return origin
        if Js==1 or Js==V.shape[1]:
            print('find_saddle: Saddle out of bounds in  y (j) direction.\n')
            return origin
    return origin

def load_soln(file):
    with open(file) as f:
        lmid = f.read()
        lmid = lmid.split('\n')[0:168]
        lmid = np.asarray(lmid)
        lmid = lmid.astype(np.float)
        l1 = lmid.astype(np.float)
    l1[147:168] = lmid[63:84]
    l1[105:126] = lmid[84:105]
    l1[63:84] = lmid[105:126]
    l1[84:105] = lmid[126:147]
    l1[126:147] = lmid[147:168]
    return l1


def write_pickle(fin,fout,grid):
    #grid is the field grid pts that give the locations of each simulated potential point
    #fin is the filename of the of the input vtk sim file
    #fout is the filename of the pickle you want to save to
    x, y, z = grid.to_xyz()
    nx = len(x)
    ny = len(y)
    nz = len(z)
    ntotal = nx * ny * nz

    trap = {'X': x,
            'Y': y,
            'Z': z}
    i = 0
    strs = "DC1 DC2 DC3 DC4 DC5 DC6 DC7 DC8 DC9 DC10 DC11 DC12 DC13 DC14 DC15 DC16 DC17 DC18 DC19 DC20 DC21".split()
    result0 = Result.load(fin, 'DC1', trans_file)
    p0 = result0.pobjectotential
    for ele in strs:
        # if ele not in excl:
        result = Result.load(fin, ele,trans_file)
        p = result.potential
        p = np.swapaxes(p, 0, 2)
        p = np.swapaxes(p, 0, 1)
        trap[ele] = {'potential': p}
        trap[ele]['position'] = [0, i]
        # else:
        #     trap[ele] = {'potential': np.zeros(np.shape(p0))}
        #     trap[ele]['position'] = [0, i]
        # i = i + 1

    electrode_list = strs

    f = open('./'+fout+'.pkl', 'wb')
    trap1 = {'X': trap['Y'],
             'Y': trap['Z'],
             'Z': trap['X'],
             'electrodes': {}}
    for electrode in electrode_list:
        trap1['electrodes'][electrode] = trap[electrode]
    pickle.dump(trap1, f, -1)
    f.close()
